src/main/java/com/nvidia/spark/rapids/jni/kudo/KudoTableHeader.java

```diff
@@ -33,37 +34,61 @@ public final class KudoTableHeader {
   /**
    * Magic number "KUD0" in ASCII.
    */
-  private static final int SER_FORMAT_MAGIC_NUMBER = 0x4B554400;
+  private static final int SER_FORMAT_MAGIC_NUMBER = 0x4B554430;
 
   // The offset in the original table where row starts. For example, if we want to serialize rows [3, 9) of the
   // original table, offset would be 3, and numRows would be 6.
-  private int offset;
-  private int numRows;
-  private int validityBufferLen;
-  private int offsetBufferLen;
-  private int totalDataLen;
-  private int numColumns;
+  private final int offset;
+  private final int numRows;
+  private final int validityBufferLen;
+  private final int offsetBufferLen;
+  private final int totalDataLen;
+  private final int numColumns;
   // A bit set to indicate if a column has a validity buffer or not. Each column is represented by a single bit.
-  private byte[] hasValidityBuffer;
+  private final byte[] hasValidityBuffer;
 
-  private boolean initialized = false;
+  /**
+   * Reads the table header from the given input stream.
+   * @param din input stream
+   * @return the table header
+   * @throws IOException if an I/O error occurs
+   */
+  public static KudoTableHeader readFrom(DataInputStream din) throws IOException {
+    int num = din.readInt();
+    if (num != SER_FORMAT_MAGIC_NUMBER) {
+      throw new IllegalStateException("Kudo format error, expected magic number " + SER_FORMAT_MAGIC_NUMBER +
+          " found " + num);
+    }
 
+    int offset = din.readInt();
+    int numRows = din.readInt();
+
+    int validityBufferLen = din.readInt();
+    int offsetBufferLen = din.readInt();
+    int totalDataLen = din.readInt();
+    int numColumns = din.readInt();
+    int validityBufferLength = lengthOfHasValidityBuffer(numColumns);
+    byte[] hasValidityBuffer = new byte[validityBufferLength];
+    din.readFully(hasValidityBuffer);
 
-  public KudoTableHeader(DataInputStream din) throws IOException {
-    readFrom(din);
+    return new KudoTableHeader(offset, numRows, validityBufferLen, offsetBufferLen, totalDataLen, numColumns,
+        hasValidityBuffer);
   }
 
   KudoTableHeader(long offset, long numRows, long validityBufferLen, long offsetBufferLen,
```

Comment:
Still confused why most of these are longs if we are going to make sure that they fit in the range of ints. Especially if this is a package private API that we are the consumers/producers of. That means that `safeLongToNonNegativeInt`, really would just become something to ensure that the value is not-negative.